doc/source/dev/depending_on_numpy.rst

```diff
@@ -167,3 +173,94 @@ alongside other, newer packages
+
+
+NumPy 2.0-specific advice
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+NumPy 2.0 is an ABI-breaking release, however it does contain support for
+building wheels that work on both 2.0 and 1.xx releases. It's important to understand that:
+
+1. When you build wheels for your package using a NumPy 1.xx version at build
+   time, those **will not work** with NumPy 2.0.
+2. When you build wheels for your package using a NumPy 2.x version at build
+   time, those **will work** with NumPy 1.xx.
+
+The first time the NumPy ABI for 2.0 is guaranteed to be stable will be the
+release of the first release candidate for 2.0 (i.e., 2.0.0rc1). Our advice for
+handling your dependency on NumPy is as follows:
+
+1. In the main (development) branch of your package, do not add any constraints.
+2. If you rely on the NumPy C API (e.g. via direct use in C/C++, or via Cython
+   code that uses NumPy), add a ``numpy<2.0`` requirement in your
+   package's dependency metadata for releases / in release branches. Do this
+   until numpy ``2.0.0rc1`` is released and you can target that.
+   *Rationale: the NumPy C ABI will change in 2.0, so any compiled extension
+   modules that rely on NumPy will break; they need to be recompiled.*
+3. If you rely on a large API surface from NumPy's Python API, also consider
+   adding the same ``numpy<2.0`` requirement to your metadata until you are
+   sure your code is updated for changes in 2.0 (i.e., when you've tested
+   things work against ``2.0.0rc1``).
+   *Rationale: we will do a significant API cleanup, with many aliases and
+   deprecated/non-recommended objects being removed (see, e.g.,*
+   :ref:`numpy-2-migration-guide` *and* :ref:`NEP52`), *so unless you only use
+   modern/recommended functions and objects, your code is likely to require at
+   least some adjustments.*
+4. Plan to do a release of your own packages which depend on ``numpy`` shortly
+   after the first NumPy 2.0 release candidate is released (probably around 1
+   Feb 2024).
+   *Rationale: at that point, you can release packages that will work with both
+   2.0 and 1.X, and hence your own end users will not be seeing much/any
+   disruption (you want* ``pip install mypackage`` *to continue working on the
+   day NumPy 2.0 is released).*
+5. Once ``2.0.0rc1`` is available, you can adjust your metadata in
+   ``pyproject.toml`` in the way outlined below.
+
+There are two cases: you need to keep compatibility with numpy 1.xx while also
+supporting 2.0, or you are able to drop numpy 1.xx support for new releases of
+your package and support >=2.0 only. The latter is simpler, but may be more
+restrictive for your users. In that case, simply add ``numpy>=2.0`` (or
+``numpy>=2.0.0rc1``) to your build and runtime requirements and you're good to
+go. We'll focus on the "keep compatibility with 1.xx and 2.x" now, which is a
+little more involved.
+
+*Example for a package using the NumPy C API (via C/Cython/etc.) which wants to support
+NumPy 1.23.5 and up*:
+
```

Comment:
Maybe mention that this goes in `pyproject.toml`? I expect we also have a `cython>= 3.0.6` requirement for downstream projects that use it. What about meson and meson-python?

Reply:
> Maybe mention that this goes in pyproject.toml? 

That's mentioned in (5) above already.

> I expect we also have a `cython>= 3.0.6` requirement for downstream projects that use it

Do we? I'd expect older Cython versions to work too, even 0.29.x versions. NumPy itself has this constraint because (a) we dropped 0.29.x support, and (b) there was a bug fix for C11 that we needed in 3.0.6

> What about meson and meson-python?

That's unrelated; any versions of those or any other build system/backend (setuptools, CMake + scikit-build-core) should work.

Reply:
> Do we? I'd expect older Cython versions to work too, even 0.29.x versions. 

Yes and no.  I expect we can keep it working, *however* with the exception of C access to the `dtype` class.

Reply:
Ah okay. Then I'd suggest to not mention anything here. Many Cython users won't be affected by that, and those that are will find out soon enough and then upgrade to a version that works.